zunis_lib/zunis/models/flows/coupling_cells/piecewise_coupling/piecewise_quadratic.py

```diff
@@ -0,0 +1,357 @@
+                                                torch.squeeze(torch.gather(v,-1,mx+1),axis=-1),alphas), axis=-1),axis=-1)),-1)
+       
+    # Regularization: points must be strictly within the unit hypercube
+    # Use the dtype information from pytorch
+    eps = torch.finfo(out.dtype).eps
+    out = out.clamp(
+        min=eps,
+        max=1. - eps
+    )
+    return out, logj
+
+
+def piecewise_quadratic_inverse_transform(y, wv_tilde, compute_jacobian=True):
+    """
+    Apply the inverse of an element-wise piecewise-linear transformation to some variables
+
+    Parameters
+    ----------
+    y : torch.Tensor
+        a tensor with shape (N,k) where N is the batch dimension while k is the
+        dimension of the variable space. This variable span the k-dimensional unit
+        hypercube
+
+    wv_tilde: torch.Tensor
+        is a tensor with shape (N,k,2b+1) where b is the number of bins.
+        This contains the un-normalized widths and heights of the bins of the piecewise-constant PDF for dimension k,
+        i.e. q_tilde lives in all of R and we don't impose a constraint on their sum yet.
+        Normalization is imposed in this function using a modified softmax.
+        
+
+    compute_jacobian : bool, optional
+        determines whether the jacobian should be compute or None is returned
+
+    Returns
+    -------
+    tuple of torch.Tensor
+        pair `(x,j)`.
+        - `x` is a tensor with shape (N,k) living in the k-dimensional unit hypercube
+        - `j` is the jacobian of the transformation with shape (N,) if compute_jacobian==True, else None.
+    """
+    
+    logj = None
+
+    # TODO do a bottom-up assesment of how we handle the differentiability of variables
+    
+    v_tilde=wv_tilde[:,:,:int(np.ceil(wv_tilde.shape[2]/2))]
+    w_tilde=wv_tilde[:,:,v_tilde.shape[2]:]
+    N, k, b = w_tilde.shape
+    
+    Nx, kx = y.shape
+    assert N == Nx and k == kx, "Shape mismatch"
+    
+    w=torch.exp(w_tilde)
+    wsum = torch.cumsum(w, axis=-1) 
+    wnorms = torch.unsqueeze(wsum[:, :, -1], axis=-1) 
+    w = w/wnorms
+    wsum=wsum/wnorms
+    wsum_shift=torch.cat((torch.zeros([wsum.shape[0],wsum.shape[1],1]).to(wsum.device, wsum.dtype),wsum),axis=-1)
+    
+    v=modified_softmax(v_tilde, w)
+    
+    #need to find the bin number for each of the y/x
+    #-> find the last bin such that y is greater than the constant of the quadratic equation
+    
+    #vw_i= (v_i+1 - v_i)w_i/2 where i is the bin index
+    vw=torch.cat((torch.zeros([v.shape[0],v.shape[1],1]).to(wsum.device, wsum.dtype),
+                                  torch.cumsum(torch.mul((v[:,:,:-1]+v[:,:,1:])/2,w),axis=-1)),axis=-1)
+    
+    finder=torch.where(vw>torch.unsqueeze(y,axis=-1),torch.zeros_like(vw),torch.ones_like(vw))
+    
+    eps = torch.finfo(vw.dtype).eps
+    mx=torch.unsqueeze(torch.argmax(torch.cat((torch.ones([vw.shape[0],vw.shape[1],1]).to(vw.device, vw.dtype)*eps,finder*(vw+1)),axis=-1),axis=-1),-1)-1
```

Comment:
Please add comments to describe the operations performed here, mx especially hard to parse.